third_party/move/move-vm/runtime/src/storage/ty_depth_checker.rs

```diff
@@ -0,0 +1,575 @@
+            .map_err(|err| {
+                // Extra safety: in case there is an error loading the definition of a struct,
+                // clear the pending cache.
+                self.pending_formulas.borrow_mut().clear();
+                err
+            })?;
+
+        let formulas = match &struct_definition.layout {
+            StructLayout::Single(fields) => fields
+                .iter()
+                .map(|(_, field_ty)| {
+                    self.calculate_type_depth_formula(gas_meter, traversal_context, field_ty)
+                })
+                .collect::<PartialVMResult<Vec<_>>>()?,
+            StructLayout::Variants(variants) => variants
+                .iter()
+                .flat_map(|variant| variant.1.iter().map(|(_, ty)| ty))
+                .map(|field_ty| {
+                    self.calculate_type_depth_formula(gas_meter, traversal_context, field_ty)
+                })
+                .collect::<PartialVMResult<Vec<_>>>()?,
+        };
+        let formula = DepthFormula::normalize(formulas);
+
+        // Add the formula to visited list and remove it from pending.
+        assert!(self.pending_formulas.borrow_mut().remove(idx));
+        let prev = self
+            .visited_formulas
+            .borrow_mut()
+            .insert(*idx, formula.clone());
+        if prev.is_some() {
+            self.pending_formulas.borrow_mut().clear();
+            self.visited_formulas.borrow_mut().clear();
+            let struct_name = self.get_struct_name(idx)?;
+            let msg = format!(
+                "Depth formula for struct {}::{}::{} is already cached",
+                struct_name.module.address, struct_name.module.name, struct_name.name
+            );
+            return Err(PartialVMError::new_invariant_violation(msg));
+        }
+
+        Ok(formula)
+    }
+
+    /// Calculates the depth formula of the specified [Type]. There are no constraints on the type,
+    /// it can also be a generic type parameter.
+    fn calculate_type_depth_formula(
+        &self,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        ty: &Type,
+    ) -> PartialVMResult<DepthFormula> {
+        Ok(match ty {
+            Type::Bool
+            | Type::U8
+            | Type::U64
+            | Type::U128
+            | Type::Address
+            | Type::Signer
+            | Type::U16
+            | Type::U32
+            | Type::U256 => DepthFormula::constant(1),
+            Type::Vector(ty) => {
+                let mut inner =
+                    self.calculate_type_depth_formula(gas_meter, traversal_context, ty)?;
+                inner.scale(1);
+                inner
+            },
+            Type::Reference(ty) | Type::MutableReference(ty) => {
+                let mut inner =
+                    self.calculate_type_depth_formula(gas_meter, traversal_context, ty)?;
+                inner.scale(1);
```

Comment:
this seems like the usage pattern in which case we could make scale return self?